File: crabzilla-example1/crabzilla-example1-service/src/main/java/io/github/crabzilla/example1/customer/CustomerModule.java
package io.github.crabzilla.example1.customer;

import com.google.inject.AbstractModule;
import com.google.inject.Provides;
import com.google.inject.Singleton;
import com.google.inject.TypeLiteral;
import com.google.inject.multibindings.MapBinder;
import io.github.crabzilla.vertx.entity.EntityCommandHandlerVerticle;
import io.github.crabzilla.vertx.entity.EntityCommandHttpRpcVerticle;
import io.vertx.core.Verticle;
import io.vertx.core.json.JsonObject;

// tag::module[]
public class CustomerModule extends AbstractModule {

  @Override
  protected void configure() {

    // to bind aggregate functions
    bind(CustomerFactory.class).asEagerSingleton();

    MapBinder<String, Verticle> mapbinder =
            MapBinder.newMapBinder(binder(), String.class, Verticle.class);

    // to bind verticles for this aggregate
    TypeLiteral<EntityCommandHttpRpcVerticle<Customer>> restType =
            new TypeLiteral<EntityCommandHttpRpcVerticle<Customer>>() {};
    TypeLiteral<EntityCommandHandlerVerticle<Customer>> handlerType =
            new TypeLiteral<EntityCommandHandlerVerticle<Customer>>() {};

    mapbinder.addBinding("A-customer.handler").to(handlerType);
    mapbinder.addBinding("Z-customer.rest").to(restType);


  }

  @Provides
  @Singleton
  EntityCommandHttpRpcVerticle<Customer> restVerticle(CustomerFactory componentsFactory, JsonObject config) {
    return componentsFactory.restVerticle(config);
  }

  @Provides
  @Singleton
  EntityCommandHandlerVerticle<Customer> handler(CustomerFactory componentsFactory) {
    return componentsFactory.cmdHandlerVerticle();
  }

}
// end::module[]

File: crabzilla-example1/crabzilla-example1-service/src/main/java/io/github/crabzilla/example1/customer/CustomerModule.kt
package io.github.crabzilla.example1.customer

import dagger.Module
import dagger.Provides
import dagger.multibindings.IntoMap
import dagger.multibindings.StringKey
import io.github.crabzilla.core.DomainEvent
import io.github.crabzilla.core.entity.*
import io.github.crabzilla.example1.SampleInternalService
import io.github.crabzilla.vertx.entity.EntityCommandHandlerVerticle
import io.github.crabzilla.vertx.entity.EntityCommandRestVerticle
import io.github.crabzilla.vertx.entity.EntityUnitOfWorkRepository
import io.github.crabzilla.vertx.modules.qualifiers.WriteDatabase
import io.vertx.circuitbreaker.CircuitBreaker
import io.vertx.core.Verticle
import io.vertx.core.Vertx
import io.vertx.core.json.JsonObject
import io.vertx.ext.jdbc.JDBCClient
import net.jodah.expiringmap.ExpiringMap
import java.util.function.BiFunction
import java.util.function.Function
import java.util.function.Supplier
import javax.inject.Singleton


// tag::module[]
@Module
class CustomerModule(val vertx: Vertx, val config: JsonObject) {

  @Provides @IntoMap
  @StringKey("rest")
  fun restVerticle(uowRepository: EntityUnitOfWorkRepository): Verticle {
    return EntityCommandRestVerticle(Customer::class.java, config, uowRepository)
  }

  @Provides @IntoMap
  @StringKey("commander")
  fun handlerVerticle(supplier: Supplier<Customer>,
                      cmdHandler: (EntityCommand, Snapshot<Customer>) -> EntityCommandResult,
                      validator: (EntityCommand) -> List<String>,
                      snapshotPromoter: SnapshotPromoter<Customer>,
                      eventRepository: EntityUnitOfWorkRepository): Verticle {
    val mycache: ExpiringMap<String, Snapshot<Customer>> = ExpiringMap.create()
    val circuitBreaker = CircuitBreaker.create("command-handler-circuit-breaker", vertx)
    return EntityCommandHandlerVerticle(Customer::class.java, supplier.get(), cmdHandler, validator, snapshotPromoter,
            eventRepository, mycache, circuitBreaker)
  }

  @Provides
  @Singleton
  fun supplierFn(service: SampleInternalService): Supplier<Customer> {
    return Supplier { Customer(sampleInternalService = service) }
  }

  @Provides
  @Singleton
  fun stateTransitionFn(): (DomainEvent, Customer) -> Customer {
    return StateTransitionFn()
  }

  @Provides
  @Singleton
  fun cmdValidatorFn(): (EntityCommand) -> List<String> {
    return CommandValidatorFn()
  }

  @Provides
  @Singleton
  fun cmdHandlerFn(stateTransitionFn: (DomainEvent, Customer) -> Customer):
          (EntityCommand, Snapshot<Customer>) -> EntityCommandResult {
    val trackerFactory : (Snapshot<Customer>) -> StateTransitionsTracker<Customer> =
            { instance -> StateTransitionsTracker(instance, stateTransitionFn)
    }
    return CommandHandlerFn(trackerFactory)
  }

  @Provides
  @Singleton
  fun snapshotPromoter(stateTransitionFn: (DomainEvent, Customer) -> Customer): SnapshotPromoter<Customer> {
    return SnapshotPromoter { instance -> StateTransitionsTracker(instance, stateTransitionFn) }
  }


  @Provides
  @Singleton
  fun customerRepo(@WriteDatabase jdbcClient: JDBCClient): EntityUnitOfWorkRepository {
    return EntityUnitOfWorkRepository(Customer::class.java, jdbcClient)
  }


}
// end::module[]